### PR TITLE
RavenDB-11747 WIP - issue not finished, has some of the fixes [4.1]

### PR DESCRIPTION
**Pull request by myarichuk**
Multiple small fixes to different issues.

### CODE REVIEW

File: src/Raven.Client/Exceptions/Database/DatabaseDoesNotExistException.cs [hunk 1]
```diff
@@ -11,18 +11,22 @@ namespace Raven.Client.Exceptions.Database
 {
     public class DatabaseDoesNotExistException : RavenException
     {
+        public string ThownAtStackTrace { get; }
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Database/DatabaseDoesNotExistException.cs]: This looks like a debugging code. Should be removed.

File: src/Raven.Server/Rachis/Leader.cs [hunk 2]
```diff
@@ -934,6 +934,8 @@ public bool TryModifyTopology(string nodeTag, string nodeUrl, TopologyModificati
                             newVotes[nodeTag] = nodeUrl;
                             break;
                         case TopologyModification.Promotable:
+                            if(nodeUrl == null)
+                                Debugger.Break();
```
Comment (ppekrol) [inline src/Raven.Server/Rachis/Leader.cs]: Debugging code.

File: src/Sparrow/Json/JsonOperationContext.cs [hunk 3]
```diff
@@ -925,7 +944,7 @@ protected internal virtual unsafe void Reset(bool forceReleaseLongLivedAllocator
                 }
 
                 _arenaAllocatorForLongLivedValues = null;
-
+                _lastResestStack = Environment.StackTrace;
```
Comment (ppekrol) [inline src/Sparrow/Json/JsonOperationContext.cs]: Debugging code. Should be removed.

File: src/Sparrow/Json/JsonOperationContext.cs [hunk 4]
```diff
@@ -890,16 +903,22 @@ private void ThrowObjectDisposed()
             throw new ObjectDisposedException(nameof(JsonOperationContext));
         }
 
+        private readonly object _renewSyncObj = new object();
         protected internal virtual void Renew()
         {
-            _arenaAllocator.RenewArena();
-            if (_arenaAllocatorForLongLivedValues == null)
+            lock (_renewSyncObj)
```
Comment (ppekrol) [inline src/Sparrow/Json/JsonOperationContext.cs]: Are we expecting to invoke this method in a concurrent manner?

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 5]
```diff
@@ -1913,8 +1913,11 @@ private async Task<(long Index, object Result)> SendToLeaderAsyncInternal(Comman
                         continue;
                     }
 
-                    var result = await SendToNodeAsync(cachedLeaderTag, cmd, reachedLeader);
-                    return (result.Index, cmd.FromRemote(result.Result));
+                    using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                    {                        
+                        var response = await SendToNodeAsync(context, cachedLeaderTag, cmd, reachedLeader);
+                        return (response.Index, cmd.FromRemote(response.Result));
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs]: IIRC, we have commands that hold `BlittableJsonReaderObject`, but here you are disposing the context while the BJRO is still in use, no?

File: src/Sparrow/Json/JsonOperationContext.cs [hunk 6]
```diff
@@ -421,13 +421,26 @@ public AllocatedMemoryData GetLongLivedMemory(int requestedSize)
...
+                allocatedMemory.ContextGeneration = Generation;
+                allocatedMemory.Parent = this;
 #if DEBUG
-            allocatedMemory.IsLongLived = true;
+                allocatedMemory.IsLongLived = true;
 #endif
-            return allocatedMemory;
+                return allocatedMemory;
+            }
+            catch (NullReferenceException e)
```
Comment (ayende) [inline src/Sparrow/Json/JsonOperationContext.cs]: We should never catch NREs. We need to handle this properly.

File: src/Sparrow/Json/JsonOperationContext.cs (hunk 4 above)
Comment (ayende) [reply src/Sparrow/Json/JsonOperationContext.cs]: There should never be such a case.

File: test/RachisTests/DatabaseCluster/EtlFailover.cs [hunk 7]
```diff
@@ -74,29 +74,45 @@ public async Task ReplicateFromSingleSource()
...
+                            Name = "Joe Doe"
+                        },"users/1");
+
+                        session.SaveChanges();
+                    }
 
-                    session.SaveChanges();
                 }
-                
+                catch (Exception e)
```
Comment (ayende) [inline test/RachisTests/DatabaseCluster/EtlFailover.cs]: can remove this

File: test/RachisTests/DatabaseCluster/EtlFailover.cs [hunk 8]
```diff
@@ -74,29 +74,45 @@ public async Task ReplicateFromSingleSource()
...
+                        session.Store(new User()
+                        {
+                            Name = "Joe Doe2"
+                        }, "users/2");
 
-                    session.SaveChanges();
+                        session.SaveChanges();
+                    }
+                    catch (Exception e)
+                    {
```
Comment (ayende) [inline test/RachisTests/DatabaseCluster/EtlFailover.cs]: remove this

**Review by ayende — CHANGES_REQUESTED** [review]

File: test/RachisTests/TopologyChangesTests.cs [hunk 9]
```diff
@@ -110,17 +111,22 @@ public async Task AddingRemovedNodeShouldWork()
             var clusterSize = 3;
             var leader = await CreateNetworkAndGetLeader(clusterSize);
             var follower = GetRandomFollower();
-            var url = follower.Url;
+
             var oldTag = follower.Tag;
+            var url = follower.Url;
             Assert.True(await leader.RemoveFromClusterAsync(oldTag).WaitAsync(TimeSpan.FromMilliseconds(leader.ElectionTimeout.TotalMilliseconds * 10)), "Was unable to remove node from cluster in time");
```
Comment (karmeli87) [inline test/RachisTests/TopologyChangesTests.cs:117]: I think this test is fixed here: https://github.com/ravendb/ravendb/pull/7404

**Pull request closed without merging**